### PR TITLE
Added the feature to disable MKL support of TensorFlow by environment…

### PR DESCRIPTION
**Pull request by cuixiaom**
…al variable TF_DISABLE_MKL=1

### CODE REVIEW

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR! We need a few minor changes.

My recent commit (https://github.com/tensorflow/tensorflow/commit/c63d21b0bfc534b6377b332e9d2ba2abbdb7e0eb) made changes to some of the files in this PR. Please `git pull`. The commit adds a define flag ENABLE_MKL to enable MKL at compile time, allowing building with MKL but not using MKL.

File: tensorflow/core/common_runtime/threadpool_device.cc [hunk 1]
```diff
@@ -49,6 +50,9 @@ ThreadPoolDevice::ThreadPoolDevice(const SessionOptions& options,
       allocator_(allocator),
       scoped_allocator_mgr_(new ScopedAllocatorMgr(name)) {
 #ifdef INTEL_MKL
+  // Eearly return when MKL is disabled
```
[the remark this answers is not in the document]
Comment (cuixiaom) [reply tensorflow/core/common_runtime/threadpool_device.cc]: fixed

File: tensorflow/core/common_runtime/process_util.cc [hunk 2]
```diff
@@ -56,6 +57,10 @@ int32 NumInterOpThreadsFromSessionOptions(const SessionOptions& options) {
   const int32 inter_op = options.config.inter_op_parallelism_threads();
   if (inter_op != 0) return inter_op;
 #ifdef INTEL_MKL
+  // Early return if MKL is disabled
+  if (DisableMKL())
+    return port::NumSchedulableCPUs();
```
[the remark this answers is not in the document]
Comment (cuixiaom) [reply tensorflow/core/common_runtime/process_util.cc]: changed

**Comment by cuixiaom** [comment]
I have made all changes as your review, please check if they are ok. We want to check it in ASAP.
Thanks!

File: tensorflow/core/common_runtime/threadpool_device.cc [hunk 3]
```diff
@@ -49,6 +50,9 @@ ThreadPoolDevice::ThreadPoolDevice(const SessionOptions& options,
       allocator_(allocator),
       scoped_allocator_mgr_(new ScopedAllocatorMgr(name)) {
 #ifdef INTEL_MKL
+  // Early return when MKL is disabled
+  if (DisableMKL())
+    return;
```
Comment (penpornk) [inline tensorflow/core/common_runtime/threadpool_device.cc]: Please run `clang-format` on all files. I believe this `if` would be formatted into one line. (No newline before`return`.)

File: tensorflow/core/util/util.cc [hunk 4]
```diff
@@ -120,4 +120,24 @@ string SliceDebugString(const TensorShape& shape, const int64 flat) {
...
+    if ((tf_disable_mkl != NULL) && (std::stoi(tf_disable_mkl) == 1)) {
+      VLOG(2) << "TF-MKL: Disabling MKL";
+      status = MKL_OFF;
+    } else {
+      status = MKL_ON;
+    }
+  }
+  return status == MKL_OFF ? true : false;
+}
+#endif
```
Comment (penpornk) [inline tensorflow/core/util/util.cc]: Please add `// INTEL_MKL` two spaces after `#endif`, i.e., <code>#endif&nbsp;&nbsp;// INTEL_MKL</code>

File: tensorflow/core/util/util.h [hunk 5]
```diff
@@ -56,6 +56,11 @@ string PrintMemory(const char* ptr, size_t n);
 // "tensor", "tensor[i]", "tensor[i, j]", etc.
 string SliceDebugString(const TensorShape& shape, const int64 flat);
 
+// disable MKL in runtime
+#ifdef INTEL_MKL
+bool DisableMKL();
+#endif
```
Comment (penpornk) [inline tensorflow/core/util/util.h]: Please add `// INTEL_MKL` two spaces after `#endif`, i.e., <code>#endif&nbsp;&nbsp;// INTEL_MKL</code>

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your quick changes! I only have a few more formatting comments.

**Comment by agramesh1** [comment]
@penpornk  I pushed changes as you had requested thanks. Also, regarding the CLA,  I am ok with my commits being contributed to this project.

**Review by penpornk — COMMENTED** [review]
Thank you for the changes, @agramesh1 !
@cuixiaom, please also confirm that you are okay with Ramesh's and (since the word `their` is kind of ambiguous) your commits being contributed to this project. Then I will approve the PR.

**Review by penpornk — APPROVED** [review]

**Comment by penpornk** [comment]
Thank you very much for the PR!

**Comment by penpornk** [comment]
Changed `cla` to `yes` because all the authors have confirmed their consent.

File: tensorflow/core/common_runtime/process_util.cc [hunk 6]
```diff
@@ -56,24 +57,26 @@ int32 NumInterOpThreadsFromSessionOptions(const SessionOptions& options) {
...
+  if (!DisableMKL()) {
+    // MKL library executes ops in parallel using OMP threads
+    // Set inter_op conservatively to avoid thread oversubscription that could
+    // lead to severe perf degradations and OMP resource exhaustion
+    int mkl_intra_op = 1;
 #ifdef _OPENMP
-  mkl_intra_op = omp_get_max_threads();
+    mkl_intra_op = omp_get_max_threads();
 #endif  // _OPENMP
-  CHECK_GE(mkl_intra_op, 1);
```
Comment (penpornk) [inline tensorflow/core/common_runtime/process_util.cc]: Please change `CHECK_GE` to `DCHECK_GE`.
There's a recent policy about replacing `CHECK` with `DCHECK`. Sorry I forgot about it earlier.

**Review by penpornk — CHANGES_REQUESTED** [review]

File: tensorflow/core/common_runtime/process_util.cc (hunk 6 above)
Comment (cuixiaom) [reply tensorflow/core/common_runtime/process_util.cc]: Changed.

**Comment by agramesh1** [comment]
I am ok with my commits being contributed to this project

**Review by penpornk — APPROVED** [review]
Thank you for the fix! :)

**Pull request merged**